core/src/network/handler/mod.rs

```diff
@@ -253,53 +262,73 @@ fn get_content_aspect(
     entry_address: &Address,
     context: Arc<Context>,
 ) -> Result<EntryAspect, HolochainError> {
-    let state = context.state()
-        .ok_or_else(|| {
-            HolochainError::InitializationFailed(
-                String::from("In get_content_aspect: no state found")
-            )
-        })?;
+    let state = context.state().ok_or_else(|| {
+        HolochainError::InitializationFailed(String::from("In get_content_aspect: no state found"))
+    })?;
 
     // Optimistically look for entry in chain...
-    let maybe_chain_header = state.agent()
+    let maybe_chain_header = state
+        .agent()
         .iter_chain()
-        .find(|ref chain_header| chain_header.entry_address() == entry_address);
+        .find(|ref chain_header| chain_header.address() == *entry_address)
+        .map(|h| (h, true))
+        .or_else(|| {
+            state
+                .agent()
+                .iter_chain()
+                .find(|ref chain_header| chain_header.entry_address() == entry_address)
+                .map(|h| (h, false))
+        });
 
     // If we have found a header for the requested entry in the chain...
-    let maybe_entry_with_header = if let Some(header) = maybe_chain_header {
-        // ... we can just get the content from the chain CAS
-        Some(EntryWithHeader {
-            entry: get_entry_from_cas(&state.agent().chain_store().content_storage(), header.entry_address())?
+    let maybe_entry_with_header = match maybe_chain_header {
+        Some((header, true)) => Some(create_entry_with_header_for_header(&*state, &header)?),
+        Some((header, false)) => {
+            // ... we can just get the content from the chain CAS
+            Some(EntryWithHeader {
+                entry: get_entry_from_cas(
+                    &state.agent().chain_store().content_storage(),
+                    header.entry_address(),
```

Comment:
Since `header.entry_address` and `entry_address` are equal, is it possible to fail faster by doing this O(1) check before searching the entire chain for the full header?